app/uk/gov/hmrc/upscanuploadproxy/controllers/UploadController.scala

```diff
@@ -81,13 +81,24 @@ class UploadController @Inject()(uriGenerator: UploadUriGenerator, proxyService:
    * Otherwise send back the failure status code with a JSON body representing the error if possible.
    */
   private def handleFailure(errorAction: ErrorAction)(failure: FailureResponse): Result = {
-    val result = errorAction.redirectUrl.fold(
-      ifEmpty = Response.json(failure.statusCode, body = XmlErrorResponse.toJson(errorAction.key, failure.body))) {
-      redirectUrl => Response.redirect(redirectUrl, queryParams = XmlErrorResponse.toFields(errorAction.key, failure.body))
+    val xmlFields = XmlErrorResponse.toFields(errorAction.key, failure.body)
+    val result =  (errorAction.redirectUrl, isErrorRedirectUrlPolicyError(xmlFields, failure.statusCode)) match {
```

Comment:
I think this tuple makes the code quite difficult to read. Might be more readable if you keep the fold as before.

```
redirectUrl => 
val xmlFields = XmlErrorResponse.toFields(errorAction.key, failure.body)
if (!isErrorRedirectUrlPolicyError(xmlFields)) {
  Response.redirect(redirectUrl, queryParams = xmlFields))
}
else {
Response.json(failure.statusCode, body = XmlErrorResponse.toJson(errorAction.key, failure.body)
}

```

This line always happens now which is probably not ideal:
```
 val xmlFields = XmlErrorResponse.toFields(errorAction.key, failure.body)
 ```
 
 But I guess another option might be a simpler pattern match
 
 ```
 case None => json
 case Some(_) if isErrorRedirectUrlPolicyError(xmlFields, failure.statusCode) => json
 case Some(redirectUrl) => // query parameter
 ```

Reply:
yea, I think that makes it simpler. 

The `xmlFields`  is calculated early but I guess in most of the cases it'll be used as the clients will provide an error redirect url.

Reply:
We have clients that are not supplying it now too. We have updated the V2 api so it can do everything that the V1 api can do.